newsroom/history_async.py

```diff
@@ -0,0 +1,178 @@
+from superdesk.flask import abort
+from superdesk.core import json
+from superdesk.core.resources.fields import ObjectId as ObjectIdField
+from eve.utils import ParsedRequest
+from superdesk.core.resources.validators import validate_data_relation_async
+from superdesk.core.resources import (
+    ResourceConfig,
+    MongoResourceConfig,
+    MongoIndexOptions,
+)
+
+
+class RouteParams(BaseModel):
+    item: Optional[Dict[str, Any]] = None
+    action: Optional[str] = None
+    section: Optional[str] = None
+
+
+class HistoryResourceModel(NewshubResourceModel):
+    action: Optional[str] = None
+    versioncreated: Optional[datetime] = None
+    user: Optional[Annotated[ObjectIdField, validate_data_relation_async("users")]] = None
+    company: Optional[Annotated[ObjectIdField, validate_data_relation_async("companies")]] = None
+    item: Optional[str] = None
+    version: Optional[str] = None
+    section: Optional[str] = None
+    extra_data: Optional[Dict] = None
+
+
+class HistoryService(NewshubAsyncResourceService[HistoryResourceModel]):
+    async def create(
+        self, docs: List[Dict[str, Any]], action: str, user: UserResourceModel, section: str = "wire", **kwargs: Any
+    ):
+        now = utcnow()
+
+        async def transform(item: Dict[str, Any]) -> Dict[str, Any]:
+            return {
+                "_id": ObjectId(),
+                "action": action,
+                "versioncreated": now,
+                "user": user.id,
+                "company": user.company,
+                "item": item["_id"],
+                "version": str(item.get("version", item.get("_current_version"))),
+                "section": section,
+            }
+
+        for doc in docs:
+            try:
+                await super().create([await transform(doc)])
+            except (werkzeug.exceptions.Conflict, pymongo.errors.BulkWriteError):
+                continue
+
+    async def create_history_record(
+        self, items: List[Dict[str, Any]], action: str, user: UserResourceModel, section: str
+    ):
+        await self.create(items, action, user, section)
+
+    async def query_items(self, query: Dict[str, Any]):
+        if query["from"] >= 1000:
+            # https://www.elastic.co/guide/en/elasticsearch/guide/current/pagination.html#pagination
+            return abort(400)
+
+        req = ParsedRequest()
+        req.args = {"source": json.dumps(query)}
+        return await super().get(req, None)
+
+    async def fetch_history(self, query: Dict[str, Any], all: bool = False):
+        results = await self.query_items(query)
+        docs = results.docs
+        if all:
+            while results.hits["hits"]["total"]["value"] > len(docs):
+                query["from"] = len(docs)
+                results = await self.query_items(query)
+                docs.extend(results.docs)
+
+        return {"_items": docs, "hits": results.hits}
+
+
+async def get_history_users(
+    item_ids: List[Union[ObjectId, str]],
```

Comment:
We can directly use Python type system here.

So:
```python
List[Union[ObjectId, str]]
```
Would become
```python
list[ObjectId | str]
```

Which is easier to read and doesn't require to import from the typing module.
Note: This only works in Python 3.10+